kr8s/_portforward.py

```diff
@@ -130,21 +146,34 @@ async def run_forever(self) -> None:
         """
         async with self:
             with contextlib.suppress(asyncio.CancelledError):
-                await self.server.serve_forever()
+                for server in self.servers:
+                    await server.serve_forever()
 
     @asynccontextmanager
     async def _run(self) -> int:
-        """Start the port forward and yield the local port."""
-        self.server = await asyncio.start_server(
-            self._sync_sockets, port=self.local_port, host="0.0.0.0"
-        )
-        async with self.server:
-            await self.server.start_serving()
-            for sock in self.server.sockets:
-                if sock.family == socket.AF_INET:
-                    yield sock.getsockname()[1]
-            self.server.close()
-            await self.server.wait_closed()
+        """Start the port forward for multiple bind addresses and yield the local port."""
+        if self.local_port == 0:
+            self.local_port = self._find_available_port()
+
+        for address in self.address:
+            server = await asyncio.start_server(
+                self._sync_sockets, port=self.local_port, host=address
+            )
+            self.servers.append(server)
+
+        try:
+            for server in self.servers:
+                async with server:
+                    await server.start_serving()
+                    for sock in server.sockets:
+                        if sock.family == socket.AF_INET:
+                            yield sock.getsockname()[1]
```

Comment:
I think this is the last thing that needs fixing up here. We can't yield inside the for loop like this. 

Now that we are choosing a random port ahead of time we don't need to infer it from the socket. So we can simplify this to:

```suggestion
            for server in self.servers:
                async with server:
                    await server.start_serving()
            yield self.local_port
```

Reply:
test fails when i just yield one time
``
        try:
            for server in self.servers:
                async with server:
                    await server.start_serving()
                    yield self.local_port
```
this code makes the tests ok

Reply:
I think the tests are passing because `127.0.0.2` and `127.0.0.3` are magic loopback addresses. With the current code only the first server gets started and the first yield happens, but then the rest never get called. The failing tests are highlighting the real bug.

I think a better test would be to start the server on `127.0.0.1` and also the IP of the machine, then test that you can connect to both.

Reply:
I have conducted local tests and confirmed that the system works as expected on 127.0.0.2 and 127.0.0.3, but it does not function correctly on 127.0.0.1, 127.0.0.4, etc. Additionally, it operates successfully on my virtual network from VMs when I configure multiple addresses not only loopback addresses.
The tests also failed in previous port forwarding scenarios when only 127.0.0.1 is set as the default address.

Reply:
Thanks for looking into this. We can't yield multiple times in the context manager, so there must be some other problem if we drop the indentation down on the yield.

Reply:
I figured out why this:
```
for server in self.servers:
                async with server:
                    await server.start_serving()
            yield self.local_port
```
does not work.
we need too yield inside in the context ```with server: ``` if it's not done server is closed before the yield. I have removed context with server: and it seems to work, the server is closed in finally block. what do you think about that ?

```
        for address in self.address:
            server = await asyncio.start_server(
                self._sync_sockets, port=self.local_port, host=address
            )
            self.servers.append(server)

        try:
            for server in self.servers:
                await server.start_serving()
            yield self.local_port


        finally:
            # Ensure all servers are closed properly
            for server in self.servers:
                server.close()
                await server.wait_closed()
                self.servers.remove(server)
``